src/circuit/gadgets/constraints.rs

```diff
@@ -698,6 +799,172 @@ mod tests {
+                let a_num = AllocatedNum::alloc(cs.namespace(|| "a_num"), || Ok(a)).unwrap();
+                let b_num = AllocatedNum::alloc(cs.namespace(|| "b_num"), || Ok(b)).unwrap();
+                let _ = enforce_equal(&mut cs, || "enforce equal", &a_num, &b_num);
+                assert_eq!(cs.is_satisfied(), a==b);
+            };
+
+            // positive
+            test_a_b(a.0, a.0);
+
+            // negative
+            test_a_b(a.0, b.0);
+        }
+
+        #[test]
+        fn test_enforce_equal_zero(a in (1u64..42)) {
+
+            let test_num = |n: u64| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let num = AllocatedNum::alloc(cs.namespace(|| "zero"), || Ok(Fr::from(n))).unwrap();
+                let _ = enforce_equal_zero(&mut cs, || "enforce equal zero", &num);
+                assert_eq!(cs.is_satisfied(), n==0);
+
+            };
+
+            // positive
+            test_num(0);
+
+            // negative
+            test_num(a);
+        }
+
+        #[test]
+        fn test_implies_equal_zero(p in any::<bool>()) {
+
+            let test_premise_num = |premise: bool, n, result: bool| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let num = AllocatedNum::alloc(cs.namespace(|| "num"), || Ok(Fr::from(n))).unwrap();
+                let pb = Boolean::constant(premise);
+                let _ = implies_equal_zero(&mut cs.namespace(|| "implies equal zero"), &pb, &num);
+                assert_eq!(cs.is_satisfied(), result);
+
+            };
+
+            // any premise
+            test_premise_num(p, 0, true);
+
+            // false premise, any value
+            let rand_a = rand::thread_rng().gen_range(0..u64::MAX);
+            test_premise_num(false, rand_a, true);
+
+            // true premise, bad values
+            let rand_positive = rand::thread_rng().gen_range(1..u64::MAX);
+            test_premise_num(true, rand_positive, false);
+        }
+
+        #[test]
+        fn test_implies_equal(p in any::<bool>(), (a, b) in any::<(FWrap<Fr>, FWrap<Fr>)>()) {
+            prop_assume!(a != b);
+
+            let test_a_b = |premise: bool, a, b, result: bool| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let a_num = AllocatedNum::alloc(cs.namespace(|| "a_num"), || Ok(Fr::from(a))).unwrap();
+                let b_num = AllocatedNum::alloc(cs.namespace(|| "b_num"), || Ok(Fr::from(b))).unwrap();
+                let pb = Boolean::constant(premise);
+                let _ = implies_equal(&mut cs.namespace(|| "implies equal"), &pb, &a_num, &b_num);
+                assert_eq!(cs.is_satisfied(), result);
+            };
+
+            // any premise
+            test_a_b(p, a.0, a.0, true);
+
+            // positive case
+            test_a_b(false, a.0, b.0, true);
+
+            // negative case
+            test_a_b(true, a.0, b.0, false);
```

Comment:
Can we make `a` and `b` be `Fr` instead of `FWrap` and avoid the noise of the `.0`? Or is the `FWrap` required so we can define strategies?

Reply:
I guess it is required, but honestly I am just mimicking previously existing unit tests in the same file, where `FWrap` was being used.

Reply:
Consider my voluminous comments an artifact of us now starting to use proptest in earnest for new work, rather than a comment on these tests in particular.